pkg/listwatch/listwatch.go

```diff
@@ -134,7 +134,9 @@ func (mlw *multiListerWatcher) List(options metav1.ListOptions) (runtime.Object,
 	defer mlw.lock.Unlock()
 
 	if mlw.state == watching || mlw.state == stopping {
+		mlw.lock.Unlock()
 		mlw.Stop()
+		mlw.lock.Lock()
```

Comment:
Hmm, why not split `Stop()` into two functions - a `StopInternal()` which assumes the lock is already held, and `Stop()`, which acquires the lock, calls `StopInternal()` and the releases the lock.  This would allow you to call `StopInternal()` here and not have to release the lock and re-acquire it (twice). 

What  happens in something else acquires the lock in between - is there a possible bug hiding in that scenario?

Reply:
Good point; that way no other goroutine can slip through the cracks

Reply:
Done. Thank you!